### ISSUE TITLE
Is it possible to improve the analyzer for `captains-log` ?

### ISSUE THREAD

**Issue by jlirochon** [exercism/elixir-analyzer#191]
On my [3rd iteration](https://exercism.org/tracks/elixir/exercises/captains-log/solutions/jlirochon) of this exercise, the analyzer gives this advice:

> One purpose of this exercise is to teach how to use Erlang libraries. Format the date using the `:io_lib` module from Erlang.

But as [you can see](https://exercism.org/tracks/elixir/exercises/captains-log/solutions/jlirochon) I used `:io_lib.format/2` already. I think what confuses the analyzer is that I have 2 bodies for `format_stardate/1`.

Is it possible to improve this ? I will happily fix it if you can give me some hints.

By the way, I had 2 bodies for `format_stardate/1` because the tests requires an `ArgumentError` to be raised when you call `format_stardate/1`. As my 1st implementation didn't call `:io_lib.format/2`, this was not the case. I added a `is_float` guard but then a `FunctionClauseError` was raised so I added another body to catch other argument types and raise the required `ArgumentError `.

I get the point now but the overall experience was not as great as it could be. I would like to improve this, if this is reachable for me.

**Comment by angelikatyborska** [comment]
Thank you for the issue! You're right, this sounds like a bad experience that should be improved.

> I think what confuses the analyzer is that I have 2 bodies for format_stardate/1.

Right. After some short debugging, I came to the conclusion that the problem is not multiple clause functions in general, but the fact that the clause that does the desired call to `:io_lib` has a guard expression with `when`.

You can see this bug in action by adding this test case to the file `test/elixir_analyzer/exercise_test/assert_call_test.exs`. This test should pass, but it doesn't:

```elixir
  test_exercise_analysis "works when calling function has a guard expression",
                         comments: [] do
    defmodule AssertCallVerification do
      def function(y) when y > 10 do
        x = List.first([1, 2, 3])
        result = helper()
        IO.puts(result)

        private_helper() |> IO.puts()
      end

      def helper do
        :helped
      end

      defp private_helper do
        :privately_helped
      end
    end
  end
```

The bug is somewhere in the implementation in `lib/elixir_analyzer/exercise_test/assert_call/compiler.ex`. Trying to fix it might be a bit too hard for a first contribution to the project, but you're welcome to try if you feel like it! Otherwise me or @jiegillet would take care of it soon.

**Comment by jiegillet** [comment]
Oh, that's a huge oversight, I'm surprised this went through unnoticed for so long! 
I suspect that the problem is here
https://github.com/exercism/elixir-analyzer/blob/d3e3980ddbe98947d6428b29c59cbf2c2ff94f25/lib/elixir_analyzer/exercise_test/assert_call/compiler.ex#L246-L250

It doesn't return the function name when there is a `when` because `when` changes the position of the function name
```elixir
iex(26)> quote do: def foo(x), do: :a
{:def, [context: Elixir, import: Kernel],
 [{:foo, [context: Elixir], [{:x, [], Elixir}]}, [do: :a]]}
iex(27)> quote do: def foo(x) when x < 10, do: :a
{:def, [context: Elixir, import: Kernel],
 [
   {:when, [context: Elixir],
    [
      {:foo, [], [{:x, [], Elixir}]},
      {:<, [context: Elixir, import: Kernel], [{:x, [], Elixir}, 10]}
    ]},
   [do: :a]
 ]}

```

So we would need to add that case (and a bunch of tests). If you feel like giving it a try, go for it.

If you have suggestions to the tests and exercises themselves, fell free to open an issue on `exercism/elixir`.
